File: frontend/src/components/tools/open_index_gen/OpenIndexGenEngine.ts
import { Matrix, pseudoInverse } from 'ml-matrix';

export type MethodType = 'unrestricted' | 'restricted' | 'desired_gains' | 'pure_desired_gains';

export interface SimulationInput {
    method: MethodType;
    P: number[][];
    G: number[][];
    v?: number[];
    restrict_idx?: number[];
    delta?: number[];
    alpha?: number;
    alpha_proportion?: number;
    cycles?: number;
    ellipse_axes?: number[];
}

export class OpenIndexGenEngine {
    
    static calcUnrestrictedIndex(P: Matrix, G: Matrix, v: Matrix): Matrix {
        // b = P^-1 * G * v
        const P_inv = pseudoInverse(P);
        return P_inv.mmul(G).mmul(v);
    }

    static calcRestrictedIndex(P: Matrix, G: Matrix, v: Matrix, restrictIdx: number[]): Matrix {
        // b = [I - P^-1 G1 (G1' P^-1 G1)^-1 G1' P^-1] P^-1 G v
        const P_inv = pseudoInverse(P);
        const n = P.rows;
        const I = Matrix.eye(n, n);
        
        // Extract G1 (columns of G corresponding to restrictIdx)
        const G1 = G.selection(Array.from({length: G.rows}, (_, i) => i), restrictIdx);
        
        // term1 = G1' * P^-1 * G1
        const term1 = G1.transpose().mmul(P_inv).mmul(G1);
        const term1_inv = pseudoInverse(term1);
        
        // bracket = I - G1 * term1_inv * G1' * P_inv
        const bracket = I.sub(G1.mmul(term1_inv).mmul(G1.transpose()).mmul(P_inv));
        const R = P_inv.mmul(bracket);
        
        return R.mmul(G).mmul(v);
    }

    static calcDesiredGains(P: Matrix, G: Matrix, v: Matrix, delta: Matrix, alpha: number, restrictIdx: number[]): { b: Matrix, alphaMax: number } {
        const P_inv = pseudoInverse(P);
        const n = P.rows;
        const I = Matrix.eye(n, n);
        
        const G1 = G.selection(Array.from({length: G.rows}, (_, i) => i), restrictIdx);
        
        const term1 = G1.transpose().mmul(P_inv).mmul(G1);
        const term1_inv = pseudoInverse(term1);
        
        const bracket = I.sub(G1.mmul(term1_inv).mmul(G1.transpose()).mmul(P_inv));
        const R = P_inv.mmul(bracket);
        
        // alpha_max_sq_inv = delta' * term1_inv * delta
        const alpha_max_sq_inv_mat = delta.transpose().mmul(term1_inv).mmul(delta);
        const alpha_max_sq_inv = alpha_max_sq_inv_mat.get(0, 0);
        
        if (alpha_max_sq_inv <= 1e-12) {
            const b_restricted = R.mmul(G).mmul(v);
            return { b: b_restricted, alphaMax: 0.0 };
        }
        
        const alphaMax = 1.0 / Math.sqrt(alpha_max_sq_inv);
        if (alpha > alphaMax) {
            alpha = alphaMax;
        }
        
        // denom_sq = v' G' R G v
        const denom_sq_mat = v.transpose().mmul(G.transpose()).mmul(R).mmul(G).mmul(v);
        const denom_sq = denom_sq_mat.get(0, 0);
        
        let scaling_factor = 0.0;
        if (denom_sq > 1e-12) {
            scaling_factor = Math.sqrt(1 - (alpha * alpha) / (alphaMax * alphaMax)) / Math.sqrt(denom_sq);
        }
        
        // term_A = scaling_factor * (R G v)
        const term_A = R.mmul(G).mmul(v).mul(scaling_factor);
        
        // term_B = alpha * P^-1 G1 term1_inv delta
        const term_B = P_inv.mmul(G1).mmul(term1_inv).mmul(delta).mul(alpha);
        
        const b = term_A.add(term_B);
        return { b, alphaMax };
    }

    static calcPureDesiredGains(P: Matrix, G: Matrix, delta: Matrix): Matrix {
        const P_inv = pseudoInverse(P);
        const term1 = G.transpose().mmul(P_inv).mmul(G);
        const term1_inv = pseudoInverse(term1);
        
        const b = P_inv.mmul(G).mmul(term1_inv).mmul(delta);
        return b;
    }

    static simulate(inputs: SimulationInput): any {
        try {
            const P = new Matrix(inputs.P);
            const G = new Matrix(inputs.G);
            let v: Matrix | null = null;
            if (inputs.v) {
                // v should be column vector
                v = Matrix.columnVector(inputs.v);
            }
            
            const method = inputs.method || 'unrestricted';
            let b: Matrix;
            let alphaMax = 0.0;
            
            if (method === 'unrestricted') {
                if (!v) throw new Error("v is required for unrestricted index.");
                b = OpenIndexGenEngine.calcUnrestrictedIndex(P, G, v);
            } else if (method === 'restricted') {
                if (!v) throw new Error("v is required for restricted index.");
                const restrictIdx = inputs.restrict_idx || [];
                if (restrictIdx.length === 0) throw new Error("restrict_idx must be provided.");
                b = OpenIndexGenEngine.calcRestrictedIndex(P, G, v, restrictIdx);
            } else if (method === 'desired_gains') {
                if (!v) throw new Error("v is required for desired gains.");
                const restrictIdx = inputs.restrict_idx || [];
                const deltaArr = inputs.delta || [];
                if (restrictIdx.length === 0 || deltaArr.length === 0) {
                    throw new Error("restrict_idx and delta must be provided.");
                }
                const delta = Matrix.columnVector(deltaArr);
                
                let alpha = inputs.alpha || 0.0;
                if (inputs.alpha_proportion !== undefined) {
                    const P_inv = pseudoInverse(P);
                    const G1 = G.selection(Array.from({length: G.rows}, (_, i) => i), restrictIdx);
                    const term1 = G1.transpose().mmul(P_inv).mmul(G1);
                    const term1_inv = pseudoInverse(term1);
                    const alpha_max_sq_inv = delta.transpose().mmul(term1_inv).mmul(delta).get(0, 0);
                    if (alpha_max_sq_inv > 0) {
                        const computedAlphaMax = 1.0 / Math.sqrt(alpha_max_sq_inv);
                        alpha = inputs.alpha_proportion * computedAlphaMax;
                    } else {
                        alpha = 0.0;
                    }
                }
                
                const res = OpenIndexGenEngine.calcDesiredGains(P, G, v, delta, alpha, restrictIdx);
                b = res.b;
                alphaMax = res.alphaMax;
            } else if (method === 'pure_desired_gains') {
                const deltaArr = inputs.delta || [];
                if (deltaArr.length === 0) throw new Error("delta must be provided.");
                const delta = Matrix.columnVector(deltaArr);
                b = OpenIndexGenEngine.calcPureDesiredGains(P, G, delta);
            } else {
                throw new Error(`Unknown method: ${method}`);
            }
            
            // delta_G = b' * G => 1xN vector
            const delta_G = b.transpose().mmul(G);
            
            return {
                status: "success",
                method,
                alpha_max: alphaMax,
                weights: b.to1DArray(),
                predicted_genetic_change: delta_G.to1DArray()
            };
        } catch (err: any) {
            return { status: "error", message: err.message || String(err) };
        }
    }

    static generateGenupEllipse(G_arr: number[][], numPoints: number = 100): any {
        const G = new Matrix(G_arr);
        const N = G.rows;
        if (N === 2) {
            return OpenIndexGenEngine._generate2DEllipse(G, numPoints);
        }
        
        const result: Record<string, any> = {};
        for (let i = 0; i < N; i++) {
            for (let j = i + 1; j < N; j++) {
                const G2 = G.selection([i, j], [i, j]);
                result[`${i}_${j}`] = OpenIndexGenEngine._generate2DEllipse(G2, numPoints);
            }
        }
        return result;
    }

    static _generate2DEllipse(G: Matrix, numPoints: number): { x: number[], y: number[] } {
        const theta = Array.from({length: numPoints}, (_, i) => (i / (numPoints - 1)) * 2 * Math.PI);
        const x_pts: number[] = [];
        const y_pts: number[] = [];
        
        for (const t of theta) {
            const v = Matrix.columnVector([Math.cos(t), Math.sin(t)]);
            const var_I = v.transpose().mmul(G).mmul(v).get(0, 0);
            if (var_I > 1e-12) {
                const sigma_I = Math.sqrt(var_I);
                const dg = G.mmul(v).mul(1.0 / sigma_I);
                x_pts.push(dg.get(0, 0));
                y_pts.push(dg.get(1, 0));
            }
        }
        return { x: x_pts, y: y_pts };
    }

    static reverseGenupEllipse(G_arr: number[][], target_x: number, target_y: number): any {
        const G = new Matrix(G_arr);
        const target_dg = Matrix.columnVector([target_x, target_y]);
        const G_inv = pseudoInverse(G);
        
        let v_rev = G_inv.mmul(target_dg);
        const norm = Math.sqrt(v_rev.get(0,0)*v_rev.get(0,0) + v_rev.get(1,0)*v_rev.get(1,0));
        if (norm > 1e-12) {
            v_rev = v_rev.mul(1.0 / norm);
        }
        
        const b_rev = v_rev;
        const var_I = b_rev.transpose().mmul(G).mmul(b_rev).get(0,0);
        const sigma_I = Math.sqrt(var_I);
        const dg_rev = G.mmul(b_rev).mul(1.0 / sigma_I);
        
        const H = v_rev.transpose().mmul(dg_rev).get(0,0);
        
        return {
            v: v_rev.to1DArray(),
            b: b_rev.to1DArray(),
            delta_G: dg_rev.to1DArray(),
            H: H
        };
    }
}

File: backend/app/core/open_index_gen_engine.py
import numpy as np
from typing import Dict, Any, Union
from .engine_base import TeachingEngine

class OpenIndexGenEngine(TeachingEngine):
    """
    Implements multi-trait selection index theory under linear constraints
    (Brascamp, 1984).
    """

    def validate_inputs(self, P: np.ndarray, G: np.ndarray, v: np.ndarray) -> None:
        """
        Validates the dimensional conformity and properties of the input matrices.
        """
        if P.shape[0] != P.shape[1]:
            raise ValueError("Phenotypic matrix P must be square.")
        
        n_info_sources = P.shape[0]
        
        if G.shape[0] != n_info_sources:
            raise ValueError(f"Genetic matrix G must have {n_info_sources} rows (matching P).")
            
        m_traits = G.shape[1]
        
        if v is not None:
            if v.shape[0] != m_traits:
                raise ValueError(f"Economic weights vector v must have {m_traits} elements (matching columns of G).")
            
        # Check positive definiteness of P by attempting Cholesky decomposition
        try:
            np.linalg.cholesky(P)
        except np.linalg.LinAlgError:
            raise ValueError("Phenotypic matrix P is not positive definite.")

    def calc_unrestricted_index(self, P: np.ndarray, G: np.ndarray, v: np.ndarray) -> np.ndarray:
        """
        Calculates Unrestricted Selection Index (Hazel, 1943).
        b = P^-1 * G * v
        """
        P_inv = np.linalg.pinv(P)
        return P_inv @ G @ v

    def calc_restricted_index(self, P: np.ndarray, G: np.ndarray, v: np.ndarray, restrict_idx: list[int]) -> np.ndarray:
        """
        Calculates Restricted Selection Index (Kempthorne & Nordskog, 1959).
        restrict_idx: indices of columns in G to restrict genetic change to 0.
        """
        P_inv = np.linalg.pinv(P)
        n = P.shape[0]
        I = np.eye(n)
        
        G1 = G[:, restrict_idx]
        
        # Robust pseudo-inverse for (G1' * P^-1 * G1)
        term1 = G1.T @ P_inv @ G1
        term1_inv = np.linalg.pinv(term1)
        
        # R = P^-1 [I - G1 (G1' P^-1 G1)^-1 G1' P^-1]
        bracket = I - (G1 @ term1_inv @ G1.T @ P_inv)
        R = P_inv @ bracket
        
        return R @ G @ v

    def calc_desired_gains(self, P: np.ndarray, G: np.ndarray, v: np.ndarray, delta: np.ndarray, alpha: float, restrict_idx: list[int]) -> tuple[np.ndarray, float]:
        """
        Calculates Desired Genetic Gains Index (Pešek & Baker, 1969; Brascamp, 1979).
        delta: vector of target genetic changes for the restricted traits.
        alpha: proportional allocation.
        restrict_idx: indices of columns in G corresponding to the restricted traits.
        Returns the index weights and alpha_max.
        """
        P_inv = np.linalg.pinv(P)
        n = P.shape[0]
        I = np.eye(n)
        
        G1 = G[:, restrict_idx]
        
        term1 = G1.T @ P_inv @ G1
        term1_inv = np.linalg.pinv(term1)
        
        bracket = I - (G1 @ term1_inv @ G1.T @ P_inv)
        R = P_inv @ bracket
        
        # alpha_max = (delta' [G1' P^-1 G1]^-1 delta)^-1/2
        alpha_max_sq_inv = delta.T @ term1_inv @ delta
        if alpha_max_sq_inv <= 0:
            raise ValueError("Invalid target genetic changes delta, leading to negative or zero term.")
            
        alpha_max = 1.0 / np.sqrt(alpha_max_sq_inv.item())
        
        if alpha > alpha_max:
            raise ValueError(f"Alpha parameter ({alpha}) exceeds maximum attainable desired gains ({alpha_max}).")
            
        # First term scaling factor
        denom_sq = v.T @ G.T @ R @ G @ v
        if denom_sq <= 0:
            raise ValueError("Denominator for scaling factor must be positive.")
        
        scaling_factor = np.sqrt(1 - (alpha**2 / alpha_max**2)) / np.sqrt(denom_sq.item())
        
        # First term
        term_A = scaling_factor * (R @ G @ v)
        
        # Second term
        term_B = alpha * P_inv @ G1 @ term1_inv @ delta
        
        b = term_A + term_B
        return b.reshape(-1, 1), alpha_max

    def calc_pure_desired_gains(self, P: np.ndarray, G: np.ndarray, delta: np.ndarray) -> np.ndarray:
        """
        Calculates Pure Desired Genetic Gains Index (Pešek & Baker, 1969)
        without economic weights.
        b = P^-1 * G * (G' * P^-1 * G)^-1 * delta
        """
        P_inv = np.linalg.pinv(P)
        term1 = G.T @ P_inv @ G
        term1_inv = np.linalg.pinv(term1)
        
        b = P_inv @ G @ term1_inv @ delta
        return b.reshape(-1, 1)

    def simulate(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        """
        Executes the simulation.
        inputs should contain:
        - method: 'unrestricted', 'restricted', 'desired_gains'
        - P: list of lists
        - G: list of lists
        - v: list
        - restrict_idx: list of integers (optional for unrestricted)
        - delta: list (optional, for desired gains)
        - alpha: float (optional, for desired gains)
        """
        try:
            P = np.array(inputs['P'], dtype=float)
            G = np.array(inputs['G'], dtype=float)
            
            v = None
            if inputs.get('v') is not None:
                v = np.array(inputs['v'], dtype=float).reshape(-1, 1)
            
            self.validate_inputs(P, G, v)
            
            method = inputs.get('method', 'unrestricted')
            
            response = {
                "status": "success",
                "method": method
            }
            
            if method == 'unrestricted':
                b = self.calc_unrestricted_index(P, G, v)
                
            elif method == 'restricted':
                restrict_idx = inputs.get('restrict_idx', [])
                if not restrict_idx:
                    raise ValueError("restrict_idx must be provided for restricted index.")
                b = self.calc_restricted_index(P, G, v, restrict_idx)
                
            elif method == 'desired_gains':
                restrict_idx = inputs.get('restrict_idx', [])
                delta = np.array(inputs.get('delta', []), dtype=float).reshape(-1, 1)
                
                # We will first calculate alpha_max using a dummy call if needed, but wait:
                # calc_desired_gains computes alpha_max inside.
                # Let's pass alpha_proportion to calc_desired_gains if provided.
                alpha_prop = inputs.get('alpha_proportion')
                if alpha_prop is not None:
                    # We need to compute alpha_max first
                    P_inv = np.linalg.pinv(P)
                    G1 = G[:, restrict_idx]
                    term1 = G1.T @ P_inv @ G1
                    term1_inv = np.linalg.pinv(term1)
                    alpha_max_sq_inv = delta.T @ term1_inv @ delta
                    if alpha_max_sq_inv > 0:
                        computed_alpha_max = 1.0 / np.sqrt(alpha_max_sq_inv.item())
                        alpha = float(alpha_prop) * computed_alpha_max
                    else:
                        alpha = 0.0
                else:
                    alpha = float(inputs.get('alpha', 0.0))
                
                if not restrict_idx or delta.size == 0:
                    raise ValueError("restrict_idx and delta must be provided for desired gains.")
                    
                b, alpha_max = self.calc_desired_gains(P, G, v, delta, alpha, restrict_idx)
                response["alpha_max"] = alpha_max
                
            elif method == 'pure_desired_gains':
                delta = np.array(inputs.get('delta', []), dtype=float).reshape(-1, 1)
                if delta.size == 0:
                    raise ValueError("delta must be provided for pure desired gains.")
                b = self.calc_pure_desired_gains(P, G, delta)
            else:
                raise ValueError(f"Unknown method: {method}")
                
            # Predicted genetic change overall: delta_G = b' G
            delta_G = b.T @ G
            
            response["weights"] = b.flatten().tolist()
            response["predicted_genetic_change"] = delta_G.flatten().tolist()
            
            # Simulate generations if cycles provided
            cycles = int(inputs.get('cycles', 0))
            if cycles > 0:
                generation_data = []
                current_g = np.zeros_like(delta_G)
                for i in range(cycles + 1):
                    generation_data.append({
                        "generation": i,
                        "cumulative_genetic_change": current_g.flatten().tolist()
                    })
                    current_g += delta_G
                response["generation_data"] = generation_data
            
            return response
            
        except Exception as e:
            return {
                "status": "error",
                "message": str(e)
            }

    def evaluate(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        """Not yet implemented for OpenIndexGen."""
        return {"status": "not_implemented"}

